runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3396,6 +3198,455 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+   dependencies->addPostCondition(scratch1, cg->getEntryPointRegister());
+   dependencies->addPostCondition(scratch2, cg->getReturnAddressRegister());
+
+   // Now generate the call to VM Helper to report the fieldwatch.
+   TR::Instruction *call = generateDirectCall(cg, node, false /*myself*/, cg->symRefTab()->findOrCreateRuntimeHelper(helperIndex, false, false, false), dependencies);
+   call->setNeedsGCMap(0x0000FFFF);
+   call->setDependencyConditions(dependencies);
+
+   // After returning from the VM Helper, branch back to mainline code.
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, mergePointLabel);
+   // End of OOL.
+   outlinedSlowPath->swapInstructionListsWithCompilation();
+
+   cg->stopUsingRegister(scratch1);
+   cg->stopUsingRegister(scratch2);
+
+   cg->stopUsingRegister(dataBlockReg);
+   }
+
+TR::Snippet *
+J9::Z::TreeEvaluator::createDataSnippetForFieldWatch(TR::CodeGenerator *cg, void *dataBlock, TR::Node *node, bool isStatic)
+   {
+   TR::S390WritableDataSnippet *dataSnippet;
+   if (isStatic)
+      {
+      dataSnippet = static_cast<TR::S390WritableDataSnippet *>(cg->CreateConstant(node, dataBlock, sizeof(J9JITWatchedStaticFieldData), true));
+      }
+   else
+      {
+      dataSnippet = static_cast<TR::S390WritableDataSnippet *>(cg->CreateConstant(node, dataBlock, 1 << (int) (log2(sizeof(J9JITWatchedInstanceFieldData)-1)+1), true));
+      }
+   return dataSnippet;
+   }
+
+void
+J9::Z::TreeEvaluator::readWriteBarrierHelperForFieldWatch(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite)
+   {
+   bool isResolved = !node->getSymbolReference()->isUnresolved();
+   TR::LabelSymbol *mergePointLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *fieldReportLabel = generateLabelSymbol(cg);
+
+   TR::Register *fieldClassReg;
+   TR::Register *fieldClassFlags = cg->allocateRegister();
+   bool opCodeIsIndirect = node->getOpCode().isIndirect();
+
+   if (opCodeIsIndirect)
+      {
+      TR::Register *objReg = cg->evaluate(node->getFirstChild());
+      // Load the class of the instance object into fieldClassReg.
+      fieldClassReg = cg->allocateRegister();
+      TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, fieldClassReg, generateS390MemoryReference(objReg, static_cast<int32_t>(TR::Compiler->om.offsetOfObjectVftField()), cg), NULL);
+      }
+   else
+      {
+      if (isResolved)
+         {
+         fieldClassReg = cg->allocateRegister();
+         J9JITWatchedStaticFieldData *staticFieldDataBlock = reinterpret_cast<J9JITWatchedStaticFieldData *>((static_cast<TR::S390WritableDataSnippet *>(dataSnippet))->getRawData());
+         J9Class *fieldClass = staticFieldDataBlock->fieldClass;
+         TR_ASSERT_FATAL(fieldClass != NULL, "A valid J9Class must be provided for direct rdbar/wrtbar opcodes %p\n", node);
+         generateRILInstruction(cg, TR::InstOpCode::LARL, node, fieldClassReg, static_cast<void *>(fieldClass));
```

Comment:
This looks dangerous in AOT. @dsouzai can you review this as well?

I suspect we need a proper JIT snippet for this `struct` and to create proper relocations for the values which will change under AOT, and code such as this should not be loading relocatable pointers directly but rather indirectly from the snippet (as they can change from one JVM to another).

Reply:
Looks like in general, on z, the instruction that is used to generate loads under AOT is something like
```
generateRegLitRefInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, targetRegister, (uintptrj_t) value, reloKind, cond, cursor, base);
```
There aren't enough bits in the `LARL` instruction to put a 64 bit address (which is necessary for AOT since you can't guarantee that the 32-bits is enough). That said, under compressed refs, 32-bits should be enough for `J9Class` pointers, but since we dont' currently support that optimization, it needs to be 64-bit.

Again though, we could specify the relo-kind, but this would likely only work for AOT w/ SVM.